src/app/shared/document/document-detail/document-detail.component.ts

```diff
@@ -0,0 +1,401 @@
+          }
+          if (row.value !== this._documentModel.data[attr.id]) {
+            row.value = this._documentModel.data[attr.id];
+          }
+        } else {
+          row = this.getRowByCorrelationId(attr.correlationId);
+          if (row) {
+            row.remove = false;
+            row.id = attr.id;
+            if (row.name !== attr.name) {
+              row.name = attr.name;
+            }
+            if (row.value !== this._documentModel.data[attr.id]) {
+              row.value = this._documentModel.data[attr.id];
+            }
+          } else {
+            this.rows.push({ id: attr.id, name: attr.name, value: this._documentModel.data[attr.id], correlationId: attr.correlationId });
+          }
+        }
+      } else {
+        this.removeRowById(attr.id);
+      }
+    });
+  }
+
+  private getRowById(id: string): DetailRow {
+    if (id) {
+      return this.rows.find(row => row.id === id);
+    }
+
+    return null;
+  }
+
+  private getRowByCorrelationId(correlationId: string): DetailRow {
+    if (correlationId) {
+      return this.rows.find(row => row.correlationId === correlationId);
+    }
+
+    return null;
+  }
+
+  private getRowByName(name: string): DetailRow {
+    if (name) {
+      return this.rows.filter(row => !row.remove).find(row => row.name === name);
+    }
+
+    return null;
+  }
+
+  private getRemovableRowByName(name: string): DetailRow {
+    if (name) {
+      return this.rows.filter(row => row.remove).find(row => row.name === name);
+    }
+
+    return null;
+  }
+
+  private removeRowById(id: string) {
+    const idx = this.rows.indexOf(this.getRowById(id));
+    if (idx >= 0 && (this.rows[idx].remove || !this._documentModel.data[id])) {
+      this.rows.splice(idx, 1);
+    }
+  }
+
+  public getNativeDate(dateObject): number {
+    if (dateObject) {
+      return new Date(dateObject.year, dateObject.monthValue, dateObject.dayOfMonth, dateObject.hour, dateObject.minute, dateObject.second).getTime();
+    }
+    return undefined;
+  }
+
+  private alreadyInCollection(attrName: string): boolean {
+    return this.collection.attributes.findIndex(attr => attr.name === attr.name) != - 1;
+  }
+
+  private prepareUpdatedDocument(): DocumentModel {
+    let updatedDocument = Object.assign({}, this._documentModel);
+
+    let dirty = this.patchNewAttributes(updatedDocument);
+    dirty = dirty || this.patchExistingAttributes(updatedDocument);
+    dirty = dirty || this.patchAttributeRename(updatedDocument);
+    dirty = dirty || this.patchReusedAttributes(updatedDocument);
+    dirty = dirty || this.patchDeletedAttributes(updatedDocument);
+
+    return dirty ? updatedDocument : null;
+  }
+
+  private patchNewAttributes(document: DocumentModel): boolean {
+    let dirty = false;
+
+    const newData: { [attributeName: string]: any } = this.rows
+      .filter(row => isNullOrUndefined(row.id) && !isNullOrUndefined(row.name) && row.name && !this.alreadyInCollection(row.name))
```

Comment:
`!isNullOrUndefined(row.name) && row.name` is the same as `row.name` condition

Reply:
good point!